TestFoundation/TestDate.swift

```diff
@@ -123,4 +124,40 @@ class TestDate : XCTestCase {
         XCTAssertTrue(d1 <= sinceReferenceDate)
         XCTAssertTrue(d2 >= sinceReferenceDate)
     }
+    
+    func test_recreateDateComponentsFromDate() {
+        let components = DateComponents(calendar: Calendar(identifier: .gregorian),
+                                        timeZone: .current,
+                                        era: 1,
+                                        year: 2017,
+                                        month: 11,
+                                        day: 5,
+                                        hour: 20,
+                                        minute: 38,
+                                        second: 11,
+                                        nanosecond: 40)
+        guard let date = Calendar(identifier: .gregorian).date(from: components) else {
+            XCTFail()
+            return
+        }
+        let recreatedComponents = Calendar(identifier: .gregorian).dateComponents(in: .current, from: date)
+        XCTAssertEqual(recreatedComponents.era, 1)
+        XCTAssertEqual(recreatedComponents.year, 2017)
+        XCTAssertEqual(recreatedComponents.month, 11)
+        XCTAssertEqual(recreatedComponents.hour, 20)
+        XCTAssertEqual(recreatedComponents.minute, 38)
+        
+        // Nanoseconds are currently not supported by UCalendar C API, returns nil
+        // XCTAssertEqual(recreatedComponents.nanosecond, 40)
```

Comment:
Is this a function of which version of ICU we're linking against?

Reply:
@parkera it seems that 59.1 but even in the last one (60.1) there are no UCalendarDateFields for nanoseconds and quarter. The list of all available UCalendarDateFields of 60.1:

enum UCalendarDateFields {
     UCAL_ERA,
     UCAL_YEAR,
     UCAL_MONTH,
     UCAL_WEEK_OF_YEAR,
     UCAL_WEEK_OF_MONTH,
     UCAL_DATE,
     UCAL_DAY_OF_YEAR,
     UCAL_DAY_OF_WEEK,
     UCAL_DAY_OF_WEEK_IN_MONTH,
     UCAL_AM_PM,
     UCAL_HOUR,
     UCAL_HOUR_OF_DAY,
     UCAL_MINUTE,
     UCAL_SECOND,
     UCAL_MILLISECOND,
     UCAL_ZONE_OFFSET,
     UCAL_DST_OFFSET,
     UCAL_YEAR_WOY,
     UCAL_DOW_LOCAL,
     UCAL_EXTENDED_YEAR,
     UCAL_JULIAN_DAY, 
     UCAL_MILLISECONDS_IN_DAY,
     UCAL_IS_LEAP_MONTH,
       /* Do not conditionalize the following with #ifndef U_HIDE_DEPRECATED_API,
        * it is needed for layout of Calendar, DateFormat, and other objects */
     UCAL_FIELD_COUNT,
     UCAL_DAY_OF_MONTH=UCAL_DATE
   };

I'm sorry if I understood something wrong.

Reply:
I think you're correct, the quarter implementation in ICU has been a problem for some time.